DirectProgramming/DPC++FPGA/include/onchip_memory_with_cache.hpp

```diff
@@ -0,0 +1,124 @@
+#ifndef __ONCHIP_MEMORY_WITH_CACHE_HPP__
+#define __ONCHIP_MEMORY_WITH_CACHE_HPP__
+
+#include <sycl/ext/intel/ac_types/ac_int.hpp>
+
+#include "constexpr_math.hpp"   // DirectProgramming/DPC++FPGA/include
+
+namespace fpga_tools {
+
+template <typename T,             // type to store in the memory
+          size_t k_mem_depth,     // depth of the memory
+          size_t k_cache_depth    // number of elements in the cache
+         >
+class OnchipMemoryWithCache {
+ public:
+  static constexpr int kNumAddrBits = fpga_tools::CeilLog2(k_mem_depth);
+  using addr_t = ac_int<kNumAddrBits, false>;
+
+  OnchipMemoryWithCache() {
+    for (int i = 0; i < k_cache_depth; i++) {
+      cache_valid_[i] = false;
+    }
+  }
+  OnchipMemoryWithCache(T init_val) {
+    for (int i = 0; i < k_mem_depth; i++) {
+      data_[i] = init_val;
+    }
+    for (int i = 0; i < k_cache_depth; i++) {
+      cache_valid_[i] = false;
+    }
+  }
+
+  OnchipMemoryWithCache(const OnchipMemoryWithCache&) = delete;
+  OnchipMemoryWithCache& operator=(const OnchipMemoryWithCache&) = delete;
+
+  // explicitly communicate to developers that we don't want to support a
+  // square bracket operator that returns a reference, as it would allow
+  // modification of the memory without updating the cache
+  template <typename I> T& operator[](I addr) = delete;
```

Comment:
Why not implement a square-bracket operator that simply calls the `write()` function like you did for the `read()`? I think `[]` is a better API than `read()` and `write()`.